### DISCUSSION TITLE
Class with __await__ marked as "Not Awaitable"

### DISCUSSION THREAD

**Discussion by YannickAlex07** [microsoft/pyright#7032] [Q&A] [answered]
Hello there! We stumbled across a weird case with an awaitable and I want to confirm if this is actually a bug in Pyright or intended / correct behavior.

The case happened while working with the [`aiomysql`](https://github.com/aio-libs/aiomysql) package and their [`create_pool()`](https://github.com/aio-libs/aiomysql/blob/83aa96e12b1b3f2bd373f60a9c585b6e73f40f52/aiomysql/pool.py#L13)-function. I will explain, however bear with me, I have to link quite a bit of code from this package.

`create_pool()` returns a custom [`_PoolContextManager`](https://github.com/aio-libs/aiomysql/blob/83aa96e12b1b3f2bd373f60a9c585b6e73f40f52/aiomysql/utils.py#L91), which in turn is a subclass of their [`_ContextManager`](https://github.com/aio-libs/aiomysql/blob/83aa96e12b1b3f2bd373f60a9c585b6e73f40f52/aiomysql/utils.py#L30) (which in turn inherits from `Coroutine`). The `_ContextManager` defines an [`__await__`](https://github.com/aio-libs/aiomysql/blob/83aa96e12b1b3f2bd373f60a9c585b6e73f40f52/aiomysql/utils.py#L70)-method, that will execute the wrapped Coroutine (in this case [`_create_pool()`](https://github.com/aio-libs/aiomysql/blob/83aa96e12b1b3f2bd373f60a9c585b6e73f40f52/aiomysql/pool.py#L20)) and then returns the result. However the `_ContextManager` also implements the `__aenter__` and `__aexit__` ([code](https://github.com/aio-libs/aiomysql/blob/83aa96e12b1b3f2bd373f60a9c585b6e73f40f52/aiomysql/utils.py#L73-L79)), making it usable as a context manager.

In our understanding it should be possible to either use the `_PoolContextManager` as an actual context manager or just await it and get a Pool back. The documentation actually always directly awaits the `create_pool()`-call and we also use this pattern without any issues so far.

However Pyright complains about the following code with a general type issue, saying that `_PoolContextManager` is not awaitable:

```python
async def main():
    pool = await aiomysql.create_pool(...)
```

On the other side the following generic code works without issues:

```python
class Test:
    def __await__(self):
        yield 1


async def main():
    await Test()
```

Now the question, is this behavior of Pyright correct here? Did we oversee anything or is our understanding just wrong? Would appreciate it a lot if we could get some clarification on this!

**Comment by erictraut** [comment]
Which version of pyright are you using? I'm not able to repro the problem with the latest version (1.1.347). I think there was a bug that I fixed a few months ago that affected awaitables.

**Reply by YannickAlex07** [reply]
I am using the newest version of Pylance - not sure if the changelog is correct, is 1.1.338 the version that is shipped with it currently?

**Reply by YannickAlex07** [reply]
Oh yeah the logs still point to 1.1.338 - then it likely is already fixed and we just need to wait for Pylance to bump the Pyright version.

**Reply by erictraut** [reply, marked as answer]
That explains it. The version of pyright that's in the latest pylance is a couple of months out of date. The pylance team is working on a big change that prevents them from updating. They're nearly finished with the change, so you should see this fixed within the next week or two.

**Reply by YannickAlex07** [reply]
That's good to hear, thanks! On a side note though, is there a way to use a custom version of Pyright with Pylance (like from a `pyproject.toml` or something)?  We usually include Pyright as a direct dependency of our project (some of us are not working with VS Code + CI/CD requires it) and we ran into a similar issue a couple of times where an issue was already fixed in Pyright but not yet in Pylance.

**Reply by debonte** [reply]
There is no way to have Pylance use any version of Pyright other than the one it is bundled with. Please take a look at https://github.com/microsoft/pylance-release/blob/main/USING_WITH_PYRIGHT.md#pyright-version-and-pylances-underlying-pyright-version-are-different and see if that helps. If not, I'd be interested in discussing your CI setup to see if there's something more we can do to help here.

**Reply by YannickAlex07** [reply]
That is of course a bummer but not really a big deal - our CI/CD is always using the latest version and only the VS Code users might see some errors that are already fixes, which usually is not a big deal as we can just run the newest version from the CLI to validate. Thanks anyway for the help and offer, really appreciate it!
